jobs,importccl: adopt protected timestamps for IMPORT

ajwerner opened the pull request:
This PR comes in two commits. See each for details.

1) The first refactors and augments the `jobs.Registry` API to enable transactional creation of a job which will be started on the local gateway. I would love to discuss this commit itself a bit.
2) Adopt protected timestamps and demonstrate their effectiveness with a test.

Commit:
jobs: create a StartableJob.

StartableJob isn't great but it's better than what we've got now.
A StartableJob is a job which is created transactionally but also hooked in
to a `resultCh` and can be started on the local gateway after the transaction
commits.

Currently we've got two main ways that jobs are kicked off "synchronously":

 1) Registry.Run()

This doesn't currently do anything to kick off the jobs - it just waits for
all of the relevant jobs to be started. This is used in conjunction with
`planner.QueueJob` and `CreateJobWithTxn`. The idea here is that when a
transaction creates a job we want, upon commit, for it to wait for the
jobs it created to complete.

 2) Registry.CreateAndStartJob()

This method has special powers that prevents the gateway node from adopting
the job and also writes the job with a lease for the local node. It also allows
the caller to pass a resultsCh which will recieve datums from the job.

That resultCh is critical for IMPORT and BACKUP (?). Currently in both of those
cases we create and run the job on a transaction outside of the caller's txn.

This is particularly frustrating because the jobs package assigns IDs at
creation time, making it difficult to transactionally associate a job with
another resource while also retaining the ability to ensure that it starts
locally and is connected to a resultsCh. This commit addresses that.

The fact that `IMPORT` and `BACKUP` commit the job before committing the
planner's txn feels dirty. I'm not planning on mucking with that stuff here in
this PR but ultimately the work here seems to lay a possible path forward. In
particular, instead of calling `Run()` with a slice of IDs we could call it
with a slice of `StartableJob`s which may be hooked up to the channel. In that
way we could unify CreateAndStartJob() and Run() with something better.

Release note: None.

Commit:
importccl: adopt protected timestamps for IMPORT

Perhaps this only needed to be applied to IMPORT INTO but it doesn't hurt
import. This change ensures that IMPORT statements protect the spans into
which they'll be writing from being GC'd.

The change adopts the StartableJob change from the previous commit,
protects a timestamp, and then makes sure to clean up that timestamp
on success or failure.

Before this change the test added here would fail with the following error:

```
require.go:1104:
                Error Trace:    import_into_test.go:152
                Error:          Expect "pq: job 526896149330362369: cannot be reverted, manual cleanup may be required: rolling back partially completed IMPORT: cannot revert before replica GC threshold 1580866345.938283735,0" to match "error response from server: 500 Internal Server Error"
                Test:           TestProtectedTimestampsDuringImportInto
```

Release note (bug fix): IMPORT INTO jobs which are canceled or fail
can no longer get stuck in an unrecoverable state if data from the
previous state of the table had expired relative to the GC TTL.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/44739)
<!-- Reviewable:end -->

ajwerner commented:
cc @lucy-zhang this is the PR with the jobs change

spaskob reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44739)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @spaskob)

___
*[pkg/jobs/registry.go, line 286 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-Qcosk5K0LvzIMwbNR:-M-Qcosl2pR7MUp8QBpg:b-iwj2p3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e4aff4f84b985450f052341f382dbce64a664b9/pkg/jobs/registry.go#L286)):*
> ```Go
> // StartableJob is a job created with a transaction to be started later.
> // See Registry.CreateStartableJob
> type StartableJob struct {
> ```

please move to jobs.go
___
*[pkg/jobs/registry.go, line 304 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-Qd5_lDOFLEHKdZ8Z8:-M-Qd5_lDOFLEHKdZ8Z9:bfva47f) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e4aff4f84b985450f052341f382dbce64a664b9/pkg/jobs/registry.go#L304)):*
> ```Go
> 		}
> 	}()
> 	// TODO(ajwerner): Consider adding some protection against starting more than
> ```

can you add detail on how that can happen?
___
*[pkg/jobs/registry.go, line 320 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-QdXCM6_o0AUXwPOyO:-M-QdXCM6_o0AUXwPOyP:b-re6loo) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e4aff4f84b985450f052341f382dbce64a664b9/pkg/jobs/registry.go#L320)):*
> ```Go
> 
> // CreateStartableJobWithTxn creates a job to be started later, after the
> // creating txn commits. It stores writes job in the jobs table, marks it
> ```

... stores writes ... ?
___
*[pkg/jobs/registry.go, line 332 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-QdOdo9CZ1LUx68bhG:-M-QdOdo9CZ1LUx68bhH:b-1adezm) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e4aff4f84b985450f052341f382dbce64a664b9/pkg/jobs/registry.go#L332)):*
> ```Go
> 		return nil, err
> 	}
> 	j.WithTxn(nil)
> ```

please add a comment why we need this
___
*[pkg/jobs/registry.go, line 338 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-Qdem3DtuPZSdDP2lB:-M-Qdem3DtuPZSdDP2lC:bpbkhvf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e4aff4f84b985450f052341f382dbce64a664b9/pkg/jobs/registry.go#L338)):*
> ```Go
> 	}
> 	resumerCtx, cancel := r.makeCtx()
> 	if err := r.register(*j.ID(), cancel); err != nil {
> ```

if the txn is aborted the job will not be unregistered.


<!-- Sent from Reviewable.io -->

spaskob reviewed (changes requested):
> Reviewed 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44739)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @spaskob)

___
*[pkg/ccl/importccl/import_stmt.go, line 630 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-QeyaO3mmnkGOhnkoY:-M-QeyaO3mmnkGOhnkoZ:b-is6aqe) ([raw file](https://github.com/cockroachdb/cockroach/blob/d965916f753df2673bbc59fbb405872252d94e12/pkg/ccl/importccl/import_stmt.go#L630)):*
> ```
> StartableJob
> ```
why can't we do that here using using QueueJob from planner.go?


<!-- Sent from Reviewable.io -->

ajwerner commented:
cc @lucy-zhang this has the jobs change we talked about - @spaskob and I discussed offline that you should certainly merge your change that utilized `QueueJob` and `Run` as written today and then later that can be refactored. This doesn't touch that API at all, just an FYI and some reassurance that any API changes in jobs will be out of your way.

dt commented on pkg/ccl/importccl/import_stmt.go:
```diff
@@ -619,20 +623,54 @@ func importPlanHook(
+		// Prepare the protected timestamp record.
+		spans := make([]roachpb.Span, 0, len(tableDetails))
+		for i := range tableDetails {
+			spans = append(spans, tableDetails[i].Desc.TableSpan())
```
I think we only need to add spans for tables where `IsNew` is false (which could require revert) and then below we can skip if spans is empty

ajwerner reviewed (commented):
This is RFAL. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-:-M12BXzi2kiYWugWo11p:b-n4viu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @spaskob)

___
*[pkg/ccl/importccl/import_stmt.go, line 630 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-QeyaO3mmnkGOhnkoY:-M12AqCv2mMBpsg57sjJ:b8dk3ek) ([raw file](https://github.com/cockroachdb/cockroach/blob/d965916f753df2673bbc59fbb405872252d94e12/pkg/ccl/importccl/import_stmt.go#L630)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

> ```
> StartableJob
> ```
why can't we do that here using using QueueJob from planner.go?
</blockquote></details>

I think we discussed this offline. 
___
*[pkg/ccl/importccl/import_stmt.go, line 652 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-I09eJqLXuJIJd_pS7-r4-652:-M12Anp63gqAczf_9Mtf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d965916f753df2673bbc59fbb405872252d94e12/pkg/ccl/importccl/import_stmt.go#L652)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I think we only need to add spans for tables where `IsNew` is false (which could require revert) and then below we can skip if spans is empty
</blockquote></details>

Done.
___
*[pkg/jobs/registry.go, line 286 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-Qcosk5K0LvzIMwbNR:-M129_4sCfeOJviL3gl1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e4aff4f84b985450f052341f382dbce64a664b9/pkg/jobs/registry.go#L286)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

please move to jobs.go
</blockquote></details>

Done.
___
*[pkg/jobs/registry.go, line 304 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-Qd5_lDOFLEHKdZ8Z8:-M-SUGK52IazyojqdbSL:b951d6l) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e4aff4f84b985450f052341f382dbce64a664b9/pkg/jobs/registry.go#L304)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

can you add detail on how that can happen?
</blockquote></details>

If there's a programming error in the client. I added protection. 
___
*[pkg/jobs/registry.go, line 320 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-QdXCM6_o0AUXwPOyO:-M11m9q79rpUJ6uAVRXg:b-lx4fj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e4aff4f84b985450f052341f382dbce64a664b9/pkg/jobs/registry.go#L320)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

... stores writes ... ?
</blockquote></details>

Fixed. 
___
*[pkg/jobs/registry.go, line 332 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-QdOdo9CZ1LUx68bhG:-M129a8f2dn8RfWhB_-D:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e4aff4f84b985450f052341f382dbce64a664b9/pkg/jobs/registry.go#L332)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

please add a comment why we need this
</blockquote></details>

Done.
___
*[pkg/jobs/registry.go, line 338 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-Qdem3DtuPZSdDP2lB:-M11md-L59kqq1EZjehT:b-eomina) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e4aff4f84b985450f052341f382dbce64a664b9/pkg/jobs/registry.go#L338)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

if the txn is aborted the job will not be unregistered.
</blockquote></details>

Good point! I've extended the interface to deal with that case and added some testing. The contract is now that user transactions which are rolled back will need to be cleaned up using `CleanupOnRollback`. I also added unit testing.


<!-- Sent from Reviewable.io -->

ajwerner commented:
erm actually let me fix the test, must have broken something

ajwerner reviewed (commented):
Okay, RFAL now

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-:-M12i8Qv8P6By4WjXOj1:ba9yflu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @spaskob)



<!-- Sent from Reviewable.io -->

ajwerner commented:
@spaskob can you give the first commit this a look? I'd like to get this merged ASAP

spaskob commented:
ok looking now

On Thu, Feb 27, 2020 at 10:12 AM ajwerner <notifications@github.com> wrote:

> @spaskob <https://github.com/spaskob> can you give the first commit this
> a look? I'd like to get this merged ASAP
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/44739?email_source=notifications&email_token=AANL635UYHO7YVKRVLFS2PLRE7J53A5CNFSM4KQCXN6KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENEWQZY#issuecomment-592013415>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANL633QVKEM627VOFKKCG3RE7J53ANCNFSM4KQCXN6A>
> .
>

Commit:
prototutil: provide mechanism for packages to register unclonable types

The logic to detect whether a type is clonable cannot traverse into all of the
implementations of itnerfaces. This is problematic when one of the structs
underneath a oneof is unclonable. While a more general solution might be nice,
this workaround allows the addition of a UUID field to a job details struct.

Release note: None

spaskob reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44739)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @spaskob)

___
*[pkg/jobs/jobs.go, line 740 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M16RwfO92e_fNxjdzg6:-M16RwfO92e_fNxjdzg7:b564fy9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0fdbcda070b78940d30b2c300eb3bc9f5848574/pkg/jobs/jobs.go#L740)):*
> ```Go
> }
> 
> // Start will resume a resumable job. The transaction used to create the
> ```

Start will start a job; resume can be confused with `RESUME job_id`
___
*[pkg/jobs/jobs_test.go, line 1994 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M16V1Lp5dLhHnuEI7br:-M16V1Lp5dLhHnuEI7bs:b-ky75yv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0fdbcda070b78940d30b2c300eb3bc9f5848574/pkg/jobs/jobs_test.go#L1994)):*
> ```Go
> }
> 
> // TestStartableJobErrors tests that the StartableJob returns the expected
> ```

I know you want to get this merged but a good to have is to search for use of CreateAndStartJob in this test file sand see if it may be useful to add similar test or replace with a StartableJob.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44739)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @spaskob)

___
*[pkg/jobs/jobs_test.go, line 1994 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M16V1Lp5dLhHnuEI7br:-M16eqwJ-RCtimep0-My:b2fpaqt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0fdbcda070b78940d30b2c300eb3bc9f5848574/pkg/jobs/jobs_test.go#L1994)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

I know you want to get this merged but a good to have is to search for use of CreateAndStartJob in this test file sand see if it may be useful to add similar test or replace with a StartableJob.
</blockquote></details>

I don't know that I follow. CreateAndStartJob is implemented on top of StartableJob. All tests which exercise it exercise the StartableJob code. 


<!-- Sent from Reviewable.io -->

spaskob reviewed (approved):
ok great then, I missed that.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-:-M16fwes8oJytaNCWtwu:b7hrtq4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @spaskob)



<!-- Sent from Reviewable.io -->

ajwerner commented:
The remaining test failures are addressed by https://github.com/cockroachdb/cockroach/pull/45544

spaskob reviewed (approved):
LGTM for jobs code 

> Reviewed 1 of 7 files at r5, 4 of 9 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-:-M1E9DKr3ssoE3RIxIOM:b-uc3o8p)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @spaskob)

___
*[pkg/ccl/importccl/import_processor_test.go, line 611 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M1E8XQ5DyIhdxpNOOJ2:-M1E8XQ5DyIhdxpNOOJ3:bw459qb) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d32c56d9f29307eff4c1f8892db55680e96ae0c/pkg/ccl/importccl/import_processor_test.go#L611)):*
> ```
> 
> ```
___
*[pkg/ccl/importccl/import_processor_test.go, line 611 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M1E8YQ-Exawx_c1vG61:-M1E8YQ-Exawx_c1vG62:bw459qb) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d32c56d9f29307eff4c1f8892db55680e96ae0c/pkg/ccl/importccl/import_processor_test.go#L611)):*
> ```
> 
> ```
___
*[pkg/ccl/importccl/import_processor_test.go, line 612 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M1E8ZqN5HmBkoKwP27R:-M1E8ZqN5HmBkoKwP27S:b3q7y3d) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d32c56d9f29307eff4c1f8892db55680e96ae0c/pkg/ccl/importccl/import_processor_test.go#L612)):*
> ```
> 			resumer := raw.(*importResumer)
> ```


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-:-M1H91ZS0sdIBdbUqjjJ:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @spaskob)

___
*[pkg/jobs/jobs.go, line 740 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M16RwfO92e_fNxjdzg6:-M1H90Of9Td9FiRvyUrE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0fdbcda070b78940d30b2c300eb3bc9f5848574/pkg/jobs/jobs.go#L740)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

Start will start a job; resume can be confused with `RESUME job_id`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->
